openapi: Updates for data-driven node parsing and multi-file schema support

adamhawkbaldwin opened the pull request:
* Build and return a `StructuralNodeModifier` list based on path parameters found
  in the provided OpenAPI spec
* Handle multiple path variables and generate appropriate regex for data driven nodes
* Use default context for adding DDNs if not provided
* Add multi-file schema support

Signed-off-by: Adam Baldwin <adam.baldwin@stackhawk.com>

kingthorin commented on addOns/openapi/CHANGELOG.md:
```diff
@@ -3,7 +3,12 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
-## Unreleased
+## [19] - 2021-04-26
```
This shouldn’t be changed until the team seems the add-on ready for release

kingthorin commented on addOns/openapi/openapi.gradle.kts:
```diff
@@ -1,6 +1,6 @@
 import org.zaproxy.gradle.addon.AddOnStatus
 
-version = "19"
+version = "20"
```
This is changed as part of the release process

thc202 commented:
The pull request is undoing changes from upstream.

psiinon commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -353,7 +360,7 @@ private static String createExceptionMessageTargetUrl(
                 urls.add(uriBuilder.merge(definitionUriBuilder));
             } catch (IllegalArgumentException e) {
-                LOG.warn("Failed to create server URL from: {}", url, e);
+                LOG.warn("Failed to create server URL from: " + url, e);
```
Revert?

psiinon commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/ValueGenerator.java:
```diff
@@ -40,11 +40,16 @@ public String getValue(String name, String type, String defaultValue) {
             defaultValue = "";
         }
         if (coreValGen == null) {
-            LOG.debug(
```
Revert?

psiinon commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/ValueGenerator.java:
```diff
@@ -60,8 +65,17 @@ public String getValue(String name, String type, String defaultValue) {
                         Collections.<String, String>emptyMap(),
                         fieldAtts);
 
-        LOG.debug(
```
Revert?

psiinon commented:
Really good to see this change, esp with a load of unit tests :)
Looks like you've unintentionally reverted some logging as I've noted.
It would also be really good to update the help to cover this new functionality.
Thats all under https://github.com/zaproxy/zap-extensions/tree/master/addOns/openapi/src/main/javahelp/org/zaproxy/zap/extension/openapi/resources/help/contents

psiinon commented:
Also should include the functionality in https://github.com/zaproxy/zap-extensions/blob/master/addOns/openapi/CHANGELOG.md

adamhawkbaldwin commented:
> Really good to see this change, esp with a load of unit tests :)
> Looks like you've unintentionally reverted some logging as I've noted.
> It would also be really good to update the help to cover this new functionality.
> Thats all under https://github.com/zaproxy/zap-extensions/tree/master/addOns/openapi/src/main/javahelp/org/zaproxy/zap/extension/openapi/resources/help/contents

Thanks @psiinon I added the relevant documentation and added the properly formatted (I believe) changelog

thc202 commented on addOns/openapi/CHANGELOG.md:
```diff
@@ -3,6 +3,14 @@ All notable changes to this add-on will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## Unreleased
+### Changed
```
Should be added to existing section.

thc202 commented on addOns/openapi/CHANGELOG.md:
```diff
@@ -3,6 +3,14 @@ All notable changes to this add-on will be documented in this file.
+## Unreleased
+### Changed
+- Add support for detecting data-driven nodes from schema
+- Build and return a `StructuralNodeModifier` list based on path parameters found
```
I don't think this is useful to the end user.

thc202 commented on addOns/openapi/CHANGELOG.md:
```diff
@@ -3,6 +3,14 @@ All notable changes to this add-on will be documented in this file.
 
+## Unreleased
+### Changed
+- Add support for detecting data-driven nodes from schema
```
Since this is all about DDN support the following items can be nested under this one.

psiinon commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java:
```diff
@@ -176,7 +178,7 @@ private String generateJsonPrimitiveValue(Schema<?> schema) {
         }
         // Should not be reached, allOf schema is resolved by the parser
-        LOG.error("Unknown composed schema type: {}", schema);
+        LOG.error("Unknown composed schema type: " + schema);
```
Should be reverted.

adamhawkbaldwin replied:
Pattern search that I should have performed earlier revealed only one other incident of this formatting change. This should be the last.

psiinon commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApi.java:
```diff
@@ -259,16 +308,34 @@ public OpenApiResults importOpenApiDefinitionV2(
+                ctx.setDataDrivenNodes(structuralNodeModifiers);
+            } catch (Exception e) {
+                LOG.error(
+                        "Error setting data driven nodes from spec for contextId: " + contextId, e);
```
We should use `{}` for log parameters - there are other instances in this PR as well.

adamhawkbaldwin commented:
Can we get an approval for this pending test please?

thc202 commented:
Running, there are conflicts though.

adamhawkbaldwin commented:
> Running, there are conflicts though.

incoming

psiinon commented:
All checks passed now - I'm assuming this is ready for review again now :)

adamhawkbaldwin commented:
rebased

thc202 commented:
Changes unrelated to DDN support should be reverted (and done in other PR). Seems some other changes were reverted but some code was still left, unused.

adamhawkbaldwin commented:
> Changes unrelated to DDN support should be reverted (and done in other PR). Seems some other changes were reverted but some code was still left, unused.

I'm sorry, could you please be more specific on where these are? I'm not seeing unused changes lingering from a revert but am likely overlooking them somewhere.

thc202 commented:
e.g. `OpenApiAPI#handleFile`, `OpenApiAPI#checkErrors`

adamhawkbaldwin commented:
> `handleFile`

Thank you, I've made the requested changes and rebased the latest commits

adamhawkbaldwin commented:
rebased latest commits*

adamhawkbaldwin commented:
rebased latest commits*

adamhawkbaldwin commented:
bump

thc202 commented:
There are still unrelated changes to DDN support.

adamhawkbaldwin commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApiTest.java:
```diff
@@ -0,0 +1,276 @@
+        // given
+        String test = "/PetStoreJson/";
+        String defnName = "defn.json";
+        String file = "v3/multi-file/api.yaml";
```
@thc202 `multi-file/*` files are pulled in for testing functionality. This `api.yaml` enumerates the sibling files in the path as dependencies for the multi-part schema composition.

thc202 replied:
I'm aware of that, I did not ask what the change was for but stated it was unrelated to DDN. If you open a PR saying that it does X and then it does W, X, Y, and Z, I can only assume the W, Y, and Z were not included on purpose (thus me asking to revert and do those unrelated changes in other PR).

If you want to add all changes in the same PR (which is fine), please, update the PR title/description accordingly. The unrelated changes should still be split into their own commits though.

adamhawkbaldwin replied:
Thank you for the clarification on this. I've updated the changlelog to reflect the additional change that was added to help support the issues encountered while in development of the core functionality addition. Will I also need to update the PR comments to cover this?

adamhawkbaldwin replied:
Thanks for the explanation and path forward. I've pushed the latest changes as two squashed commits representing the work done for the multi-file support and the additional commit for the DDN support.

Please let me know if I've missed anything.

adamhawkbaldwin commented on addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/multi-file/api.yaml:
```diff
@@ -0,0 +1,94 @@
+  /user/logout:
+    $ref: 'user.api.yaml#/user-logout'
+  /user/{username}:
+    $ref: 'user.api.yaml#/user-username'
```
☝️ as noted in the other comment, these subsequent files are pulled in as part of a multi-part schema.

adamhawkbaldwin reviewed (commented):
@thc202 I've made comments regarding the aforementioned test files and grokked for other unused code. New commit inbound. 

If I have missed anything else may I ask that these comments be placed in the area of concern for me to review? You'll have to forgive me. I'm a little dense. Unambiguous indicators will aid me in getting this into compliance quicker and should ease the review process while attempting to merge back into the codebase. 

Thank you!

psiinon commented:
@adamhawkbaldwin are you going to split the unrelated changes into separate commits as per https://github.com/zaproxy/zap-extensions/pull/2913#discussion_r637908828 ?

adamhawkbaldwin commented:
> @adamhawkbaldwin are you going to split the unrelated changes into separate commits as per [#2913 (comment)](https://github.com/zaproxy/zap-extensions/pull/2913#discussion_r637908828) ?

@psiinon I've split the commits as per the prior instruction. It looks like there are some updates to be merged as well. In the meantime if there are any other issues on this latest push let me know and I’ll get them into my next commit. Cheers

adamhawkbaldwin commented:
bump

thc202 commented:
The DDN commit is still including changes from multi-file (e.g. test files), it's also reverting changes from the multi-file commit (they should not be done in the first place).

adamhawkbaldwin commented:
> The DDN commit is still including changes from multi-file (e.g. test files), it's also reverting changes from the multi-file commit (they should not be done in the first place).

I've cleaned up the overlap on these commits. Please let me know if I missed anything.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+        return structuralNodeModifiers;
+    }
+
+    private StructuralNodeModifier createStructuralNode(Pattern pattern, String ddnName) {
```
`static`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+
+    private String createStructuralNodeRegexString(
+            List<String> params, String finalTargetUrl, String key) {
+        if (params.size() > 0) {
```
`isEmpty` and return, easier to know what the outcome is.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+
+            StringBuilder sb = new StringBuilder();
+            if (finalTargetUrl != null) {
+                sb.insert(0, finalTargetUrl);
```
Can just append, `sb` is still empty.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+                sb.insert(0, finalTargetUrl);
+            }
+
+            sb.append("(");
```
Append char.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+                StructuralNodeModifier.Type.DataDrivenNode, pattern, ddnName);
+    }
+
+    private String createStructuralNodeRegexString(
```
`static`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+        return null;
+    }
+
+    private List<String> getPathParameters(String key) {
```
`static`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+        String[] pathParts = key.split("/");
+        List<String> params = new ArrayList<>();
+        for (String part : pathParts) {
+            if (part.matches("\\{.*}")) {
```
It could be compiled once.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +385,124 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+        return params;
+    }
+
+    private String removeBrackets(String str) {
```
`static`

thc202 commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/v3/OpenApiUnitTest.java:
```diff
@@ -473,6 +474,39 @@ void shouldGenerateContentTypeHeaders()
     }
 
+    @Test
+    public void shouldGenerateJsonRequestsBody()
```
- This test does not seem to assert.

thc202 commented on addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/PetStore_defn.json:
```diff
@@ -0,0 +1,1040 @@
+{
```
It would be better if the test definitions were reduced to the minimum required for the test (otherwise we end up having several tests testing the same thing).

adamhawkbaldwin replied:
@thc202 I've reduced test definitions to what I can determine is appropriate for new and existing cases. Please let me know if there are other optimizations you had in mind that I may have missed. Thanks.

thc202 replied:
This file can be removed and used the existing one with the same contents.

thc202 replied:
There are other models/endpoints that can be removed as well.

thc202 commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApiTest.java:
```diff
@@ -0,0 +1,261 @@
+        List<String> errors = classUnderTest.importOpenApiDefinition(tmpFile, false);
+
+        // then
+        assertThat("Should parse OK: " + errors, errors.isEmpty());
```
It's preferable to use matchers which already provide a (good) error message than manually add a reason everywhere.

thc202 commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApiTest.java:
```diff
@@ -0,0 +1,261 @@
+import org.zaproxy.zap.model.Context;
+import org.zaproxy.zap.testutils.NanoServerHandler;
+
+public class ExtensionOpenApiTest extends AbstractServerTest {
```
- Remove public modifiers.
 - Some of the tests should really be in `SwaggerConverterUnitTest`.

thc202 replied:
The latter was not yet addressed, deferring.

thc202 commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApiTest.java:
```diff
@@ -0,0 +1,261 @@
+
+public class ExtensionOpenApiTest extends AbstractServerTest {
+
+    private ExtensionOpenApi classUnderTest;
```
Probably better not introduce other naming scheme.

thc202 commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApiTest.java:
```diff
@@ -0,0 +1,261 @@
+        return ctx;
+    }
+
+    private static void setControlSingleton(Control control) throws Exception {
```
Can't this rely on `Control#initSingletonForTesting` instead?

psiinon commented:
Sorry, looks like my fix has made this conflict :/

thc202 commented:
Conflict addressed.

thc202 commented:
One of the tests is failing, will take a look at that.

thc202 commented:
The mult-file import is not checking if there are any errors while parsing the definitions, one can remove the referenced files and the tests still pass. (That's unrelated to the above test failure.)

thc202 commented:
Will check that along the test failure.

adamhawkbaldwin commented:
> The mult-file import is not checking if there are any errors while parsing the definitions, one can remove the referenced files and the tests still pass. (That's unrelated to the above test failure.)

I've fixed the error handling in the results parsing check routine.

thc202 commented on addOns/openapi/CHANGELOG.md:
```diff
@@ -4,13 +4,18 @@ All notable changes to this add-on will be documented in this file.
 - Added support for Multipart form-data (Issue 6418).
 
 ### Changed
+- Add support for detecting data-driven nodes from schema
+    - Handle multi-level path parameters and generate appropriate regex for DDNs
+    - Use default context for adding DDNs if not provided
```
Needs to be removed.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApi.java:
```diff
@@ -248,18 +275,47 @@ public void importOpenApiDefinition(final File file) {
+            if (openApi == null || !swaggerParseResult.getMessages().isEmpty()) {
+                results.setErrors(swaggerParseResult.getMessages());
+            } else {
+                importOpenApiDefinition(
```
It's now ignoring the errors that might happen while importing.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/OpenApiAPI.java:
```diff
@@ -71,42 +75,47 @@ public String getPrefix() {
     @Override
     public ApiResponse handleApiAction(String name, JSONObject params) throws ApiException {
         if (ACTION_IMPORT_FILE.equals(name)) {
-            File file = new File(params.getString(PARAM_FILE));
```
This needs to be reverted, now all errors are handled as illegal parameter for the target parameter.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/OpenApiAPI.java:
```diff
@@ -129,4 +138,24 @@ public ApiResponse handleApiAction(String name, JSONObject params) throws ApiExc
         }
     }
+
+    private Integer getContextId(JSONObject params) throws ApiException {
```
- `static`
 - `int`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/OpenApiAPI.java:
```diff
@@ -129,4 +138,24 @@ public ApiResponse handleApiAction(String name, JSONObject params) throws ApiExc
+                }
+            }
+        } catch (ApiException e) {
+            throw new ApiException(ApiException.Type.ILLEGAL_PARAMETER, PARAM_TARGET);
```
The catch should be removed, the above method already throws the expected exception.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +388,137 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+        return structuralNodeModifiers;
+    }
+
+    public List<String> setStructuralNodeModifiers(int contextId) {
```
Does not need to return the `errors` as it's already adding to the list of `errors`.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +388,137 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+    private static String createStructuralNodeRegexString(
+            List<String> params, String finalTargetUrl, String key) {
+        if (params.isEmpty()) return null;
+        else {
```
The else is unnecessary.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java:
```diff
@@ -381,4 +388,137 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
+        List<StructuralNodeModifier> structuralNodeModifiers = new ArrayList<>();
+        try {
+            OpenAPI openAPI = getOpenAPI();
+            String targetUrl;
+            if (targetUriBuilder.isEmpty()) {
+                targetUrl = null;
+            } else {
+                targetUrl = targetUriBuilder.build();
+            }
```
This should use one of the resolved API URLs (it's incorrect when the user does not specify the target URL).

thc202 commented on addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v2/multi-file/api.yaml:
```diff
@@ -0,0 +1,68 @@
+
+schemes:
+  - https
+host: www.google.com
```
Needs to use test server.

thc202 commented on addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/MultiVarPath_defn.yaml:
```diff
@@ -0,0 +1,90 @@
+  description: OpenAPI v3 specs for Thing
+  version: "0.1"
+servers:
+  - url: /
```
Should use the test server.

thc202 commented on addOns/openapi/src/test/resources/org/zaproxy/zap/extension/openapi/v3/multi-file/api.yaml:
```diff
@@ -0,0 +1,68 @@
+  description: Find out more about Swagger
+  url: http://swagger.io
+servers:
+  - url: https://petstore.swagger.io/v2
+  - url: http://petstore.swagger.io/v2
```
Should use the test server.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/ExtensionOpenApi.java:
```diff
@@ -218,48 +228,71 @@ public OpenApiResults importOpenApiDefinitionV2(
+            } else {
+                importOpenApiDefinition(
+                        Json.pretty(openApi), targetUrl, null, initViaUi, requestor, contextId);
+            }
```
This was not yet addressed, deferring for later PR.

thc202 commented:
Addressed the test failure and remaining comments.

Please let me know if you have any question/comment regarding the pushed changes.

thc202 commented:
Thank you!

psiinon commented:
Many thanks @adamhawkbaldwin - what details would like like us to use for you for the credits? https://www.zaproxy.org/docs/desktop/credits/

adamhawkbaldwin commented:
> Many thanks @adamhawkbaldwin - what details would like like us to use for you for the credits? https://www.zaproxy.org/docs/desktop/credits/

I suppose let's go with the [name (twitter_handle)] format for "Adam Baldwin (@AdamHawkBaldwin)" Thanks @psiinon
